src/codegen/exeffToNoeff.ml

```diff
@@ -0,0 +1,360 @@
+    then ( failwith "Empty match statement" )
+    else ( let ((p1,c1) :: _ ) = abs_lst in
+           let (tyc, elabc) = comp_elab state env c1 in 
+           (tyc, NoEff.NMatch (elabv, List.map elab_abs abs_lst, loc)) ) )
+  | ExEff.Apply (v1, v2) -> 
+    let (ty1, elab1) = value_elab state env v1 in
+    ( match ty1 with
+      | ExEffTypes.Arrow (t1, t2) -> 
+        let (ty2, elab2) = value_elab state env v2 in
+          if (ty2 = t1)
+          then (t2, NoEff.NApplyTerm (elab1, elab2))
+          else (failwith "Improper argument type")
+      | _ -> failwith "Improper function type" )
+  | ExEff.Handle (value, comp) -> 
+    let ( (ctype, cdirt), elabc ) = comp_elab state env comp in 
+    let ( vtype, velab ) = value_elab state env value in
+    ( match vtype with
+      | ExEffTypes.Handler ( (vty1, vdirt1), (vty2, vdirt2) ) ->
+        if (vty1 = ctype && vdirt1 = cdirt) then (
+          if (Types.is_empty_dirt cdirt)
+          (* Handle - Case 1 *)
+          then ( (vty2, vdirt2), NoEff.NApplyTerm (velab, elabc))
+          else (
+               if (Types.is_empty_dirt vdirt2)
+               (* Handle - Case 2 *)
+               then ( let (_, telab) = type_elab state env vty2 in
+                 ( (vty2, vdirt2), 
+                 NoEff.NCast (NoEff.NHandle (elabc, velab),
+                 NoEff.NCoerUnsafe (NoEff.NCoerRefl (telab)))) )
+               (* Handle - Case 3 *)
+               else ( (vty2, vdirt2), NoEff.NHandle (elabc, velab))
+               )
+        )
+        else failwith "Handler source type and handled computation type do not match"
+      | _ -> failwith "Ill-typed handler" )
+  | ExEff.Call ((eff, (ty1, ty2)), value, (p, ty, comp)) ->
+    let (_, t1) = type_elab state env ty1 in
+    let (_, t2) = type_elab state env ty2 in
+    let (_, tt) = type_elab state env ty in
+    let (vty, velab) = value_elab state env value in
+    if (vty = ty1) 
+    then (
+      let env' = extend_pattern_type env p ty in
+      let (cty, celab) = comp_elab state env' comp in
+      ( cty, NoEff.NCall ((eff, (t1, t2)), velab, (pattern_elab p, tt, celab)) )
+    )
+    else failwith "Ill-typed call"
+  | ExEff.Op ((eff, (ty1, ty2)), value) -> 
+    let (_, t1) = type_elab state env ty1 in
+    let (_, t2) = type_elab state env ty2 in
+    let (vty, velab) = value_elab state env value in    
+    if (vty = ty1)
+    then ( ((ty2, ExEffTypes.empty_dirt), NoEff.NOp ((eff, (t1,t2)), velab)) )
+    else failwith "Ill-typed operation"
+  | ExEff.Bind (c1, (p, c2)) ->
+    let ((ty1, dirt1), elab1) = comp_elab state env c1 in
+    let env' = extend_pattern_type env p ty1 in
+    let ((ty2, dirt2), elab2) = comp_elab state env' c2 in
+    if (ExEffTypes.is_empty_dirt dirt1 && ExEffTypes.is_empty_dirt dirt2)
+    (* Bind - Case 1 *)
+    then ( (ty2, dirt2), NoEff.NLet (elab1, (pattern_elab p, elab2)) )
+    (* Bind - Case 2 *)
+    else ( (ty2, dirt2), NoEff.NBind (elab1, (pattern_elab p, elab2)) )
```

Comment:
Because you are also type-checking, you should make sure that `dirt1` and `dirt2` are the same (otherwise the program is ill-typed).